materialize-slack/main.go

```diff
@@ -0,0 +1,300 @@
+		}
+		bindings = append(bindings, &binding{
+			resource:   res,
+			collection: string(b.Collection.Name),
+			spec:       b,
+		})
+	}
+
+	var transactor = &transactor{
+		api:      api,
+		bindings: bindings,
+	}
+
+	return transactor, &pm.Response_Opened{}, nil
+}
+
+type transactor struct {
+	api      *SlackAPI
+	bindings []*binding
+}
+
+type binding struct {
+	spec       *pf.MaterializationSpec_Binding
+	resource   resource
+	collection string
+}
+
+func buildDocument(b *binding, keys, values tuple.Tuple) map[string]interface{} {
+	var document = make(map[string]interface{})
+
+	// Add the keys to the document.
+	for i, value := range keys {
+		var propName = b.spec.FieldSelection.Keys[i]
+		document[propName] = value
+	}
+
+	// Add the non-keys to the document.
+	for i, value := range values {
+		var propName = b.spec.FieldSelection.Values[i]
+
+		if raw, ok := value.([]byte); ok {
+			document[propName] = json.RawMessage(raw)
+		} else {
+			document[propName] = value
+		}
+	}
+	return document
+}
+
+func (d *transactor) Load(it *pm.LoadIterator, loaded func(int, json.RawMessage) error) error {
+	for it.Next() {
+		panic("Load should never be called for materialize-slack")
+	}
+	return nil
+}
+
+func (t *transactor) Store(it *pm.StoreIterator) (pm.StartCommitFunc, error) {
+
+	for it.Next() {
+		var b = t.bindings[it.Binding]
+		var parsed = buildDocument(b, it.Key, it.Values)
+		var tsStr, tsOk = parsed["ts"].(string)
+		var text, textOk = parsed["text"].(string)
+		var blocks = parsed["blocks"].(json.RawMessage)
+
+		if !tsOk {
+			return nil, fmt.Errorf("missing timestamp")
+		}
+
+		if !textOk {
+			return nil, fmt.Errorf("missing text")
+		}
+
+		// Parse the timestamp as a time.Time
+		ts, err := time.Parse(time.RFC3339Nano, tsStr)
+		if err != nil {
+			return nil, fmt.Errorf("invalid timestamp %q", tsStr)
+		}
+
+		var blocksParsed slack.Blocks
+
+		// Is this jank? This should really be taken care of by the
+		// serialization logic in slack.Block, but it can't be _that_
+		// bad to do it here... right? see below:
+		// https://api.slack.com/reference/surfaces/formatting#escaping
+		var escaped_blocks = string(blocks)
+		escaped_blocks = strings.ReplaceAll(escaped_blocks, "&", "&amp;")
+		escaped_blocks = strings.ReplaceAll(escaped_blocks, "<", "&lt;")
+		escaped_blocks = strings.ReplaceAll(escaped_blocks, ">", "&gt;")
+
+		if err := json.Unmarshal([]byte(escaped_blocks), &blocksParsed); err != nil {
+			return nil, fmt.Errorf("invalid blocks value %q, %q: %w", reflect.TypeOf(blocks), string(blocks), err)
+		}
+
+		// Accept messages from at most 10 minutes in the past
+		if time.Since(ts).Minutes() < 10 {
+			if err := t.api.PostMessage(b.resource.Channel, text, blocksParsed.BlockSet, b.resource.SenderConfig); err != nil {
+				return nil, fmt.Errorf("error sending message: %w", err)
+			}
+			time.Sleep(time.Second * 10)
+		} else {
+			log.Warn(fmt.Sprintf("Ignoring message from the past: %q", ts))
```

Comment:
Having this be logged at anything other than `debug` is going to be super spammy if this gets pointed at a big collection of historical documents. Also consider using structured logging along the lines of `log.WithField("ts", ts).Debug("ignoring message from the past")`